### ISSUE TITLE
arithmetic operations on datetime in WHERE clause don't work

### ISSUE THREAD

**Issue by larisau** [scylladb/scylladb#3668]
Scylla version (or git commit hash): master 
https://github.com/scylladb/scylla/commit/fce10f2c6e721e5961261eca27955840d7c7bc85
Cluster size: single node

cqlsh> CREATE TABLE ks.time (k int, t timestamp, PRIMARY KEY (k, t));
cqlsh> insert INTO ks.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh> insert INTO ks.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh> insert INTO ks.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh> insert INTO ks.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh> insert INTO ks.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh> select * from ks.time ;

 k | t
---+--------------------------
 0 | 2018-08-07 13:32:40+0000
 0 | 2018-08-07 13:33:20+0000
 0 | 2018-08-07 13:33:24+0000
 0 | 2018-08-07 13:33:26+0000
 0 | 2018-08-07 13:33:28+0000

(5 rows)

cqlsh> select * from ks.time WHERE t >= '2018-08-07 13:33:28+0000' + 2s ALLOW FILTERING ;
**SyntaxException: line 1:60  : syntax error...**
cqlsh> select * from ks.time WHERE t >= toTimeStamp(now()) + 1d ALLOW FILTERING ;
**SyntaxException: line 1:52  : syntax error...**

even simple select without arithmetic didn't work:
cqlsh> select * from ks.time WHERE t = '2018-08-07 13:33:28+0000' ALLOW FILTERING ;

 k | t
---+---

(0 rows)

**Comment by tzach** [comment]
```
select * from ks.time WHERE t >= '2018-08-07 13:33:28+0000' + 2s ALLOW FILTERING ;
```

Does not work for Apache Cassandra 3.11 either

**Comment by duarten** [comment]
Shouldn't there be an issue for the select issue? Much more important than the arithmetic one.

**Comment by tzach** [comment]
@duarten good point.
@larisau could you please reproduce just the "without arithmetic" case an open a new issue for it Thanks

**Comment by larisau** [comment]
apache cassandra doc:
http://cassandra.apache.org/doc/latest/cql/operators.html

**Comment by larisau** [comment]
The case without arithmetic:
works this way:
```
cqlsh> select * from ks.time WHERE t >= '2018-08-07 13:33:28+0000' ALLOW FILTERING ;
```
and does not work this:

cqlsh> select * from ks.time WHERE t = '2018-08-07 13:33:28+0000' ALLOW FILTERING ;
```
Should it work?

**Comment by duarten** [comment]
It should work. I guess there's some representation issue going on. Maybe the underlying timestamp type stores milliseconds (although milliseconds are not displayed in your `select * from ks.time` query), but milliseconds are absent in the `t = '2018-08-07 13:33:28+0000'` restriction.

**Comment by duarten** [comment]
In cassandra trunk I get:

```
cqlsh:ks> insert INTO ks.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh:ks> select * from time;

 k | t
---+---------------------------------
 0 | 2018-08-09 10:54:33.702000+0000
```

Which shows the milliseconds. If I try to query without them:

```
cqlsh:ks> select * from ks.time WHERE t = '2018-08-09 10:54:33+0000' ALLOW FILTERING ;

 k | t
---+---

(0 rows)
```

However, I haven't found a way to query _with_ the milliseconds:

```
cqlsh:ks> select * from ks.time WHERE t >= '2018-08-09 10:54:33.702000+0000' ALLOW FILTERING ;
InvalidRequest: Error from server: code=2200 [Invalid query] message="Unable to coerce '2018-08-09 10:54:33.702000+0000' to a formatted date (long)"
```

Anyway, this probably warrants a separate issue.

**Comment by larisau** [comment]
I run it from python and used 'isoformat' converting the timestamp to string format, so the timestamp values were:
['2018-08-09T10:52:57.893000', '2018-08-09T10:52:58.895000', '2018-08-09T10:52:59.898000', '2018-08-09T10:53:00.901000', '2018-08-09T10:53:01.904000']

running query with this value got the error:
```
cqlsh> SELECT * FROM ks.time WHERE t = '2018-08-09T10:52:57.893000' ALLOW FILTERING;
InvalidRequest: Error from server: code=2200 [Invalid query] message="marshaling error: unable to parse date '2018-08-09T10:52:57.893000': marshaling error: Milliseconds length exceeds expected (6)"
```
trim zeros: no error, but still empty:
```
SELECT * FROM ks.time WHERE t = '2018-08-09T10:52:57.893' ALLOW FILTERING;
 k | t
---+---

(0 rows)

```

**Comment by duarten** [comment]
Strange. Can you open a dedicated issue when you have time?

**Comment by guy9** [comment]
Did we open a separate issue for this?
It seems like the issue is related to the +0000 part of the query (the RFC 822 4-digit time zone specification).
If I run a query with the +0000 part, I get the
 _"InvalidRequest: Error from server: code=2200 [Invalid query] message="marshaling error: unable to parse date '2023-01-02 10:31:47.982000+0000': marshaling error: Milliseconds length exceeds expected (6)"_ 
error. 
Without this part, it works:

```
cqlsh>  CREATE TABLE catalog.time (k int, t timestamp, PRIMARY KEY (k, t));
cqlsh> insert INTO catalog.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh> insert INTO catalog.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh> select * from catalog.time ;

 k | t
---+---------------------------------
 0 | 2023-01-02 10:31:45.280000+0000
 0 | 2023-01-02 10:31:47.982000+0000

(2 rows)
cqlsh> select * from catalog.time  WHERE t = '2023-01-02 10:31:47.982000+0000';
InvalidRequest: Error from server: code=2200 [Invalid query] message="marshaling error: unable to parse date '2023-01-02 10:31:47.982000+0000': marshaling error: Milliseconds length exceeds expected (6)"
cqlsh> select * from catalog.time  WHERE t = '2023-01-02 10:31:47.982';

 k | t
---+---------------------------------
 0 | 2023-01-02 10:31:47.982000+0000

(1 rows)
cqlsh> 

```

**Comment by tzach** [comment]
@guy9 what verion of Scylla are you using.

This
```
select * from catalog.time  WHERE t = '2023-01-02 10:31:47.982';
```
should not work with out ALLOW FILTERING

For reference here is the same with Apache Cassandra 3.11
```
cqlsh:mykeyspace> CREATE KEYSPACE catalog WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;
cqlsh:mykeyspace> CREATE TABLE catalog.time (k int, t timestamp, PRIMARY KEY (k, t));
cqlsh:mykeyspace> insert INTO catalog.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh:mykeyspace> insert INTO catalog.time (k, t) VALUES (0, toTimeStamp(now()));
cqlsh:mykeyspace> select * from catalog.time ;

 k | t
---+---------------------------------
 0 | 2023-01-03 06:53:47.992000+0000
 0 | 2023-01-03 06:53:52.499000+0000

(2 rows)
cqlsh:mykeyspace> select * from catalog.time  WHERE t = '2023-01-02 10:31:47.982000+0000';
InvalidRequest: Error from server: code=2200 [Invalid query] message="Unable to coerce '2023-01-02 10:31:47.982000+0000' to a formatted date (long)"
cqlsh:mykeyspace> select * from catalog.time  WHERE t = '2023-01-02 10:31:47.982';
InvalidRequest: Error from server: code=2200 [Invalid query] message="Cannot execute this query as it might involve data filtering and thus may have unpredictable performance. If you want to execute this query despite the performance unpredictability, use ALLOW FILTERING"
cqlsh:mykeyspace> 


```

**Comment by guy9** [comment]
The above was 4.5. Now I ran it in 5.1, and a warning is displayed with regards to ALLOW FILTERING:

```
cqlsh> select * from catalog.time  WHERE t = '2023-01-02 10:31:47.982';

 k | t
---+---

(0 rows)

Warnings :
This query should use ALLOW FILTERING and will be rejected in future versions.

cqlsh>
```

I checked again, and the issue does not have to do with the  +0000 part of the query (the RFC 822 4-digit time zone specification).
It seems like we expect a precision of 3 milliseconds while select all displays a precision of 6 milliseconds. 
This works:
```
cqlsh> select * from catalog.time  WHERE t = '2023-01-03 07:03:28.253' allow filtering;

 k | t
---+---------------------------------
 0 | 2023-01-03 07:03:28.253000+0000

(1 rows)
cqlsh> select * from catalog.time  WHERE t = '2023-01-03 07:03:28.253+0000' allow filtering;

 k | t
---+---------------------------------
 0 | 2023-01-03 07:03:28.253000+0000

(1 rows)
cqlsh>
```

Cassandra behaves similarly, except they give a different error message. 
In the above case, do we store a precision of 3 milliseconds or 6? If it's 3, I suggest we either:
1. Keep things as they are and document this, perhaps in "troubleshooting"
2. Change the query result to display a precision of 3 milliseconds instead of 6. This would be different from Cassandra.

**Comment by nyh** [comment]
@guy9 it's not clear to me if you are saying that this issue got fixed since it was created four years ago, or there is still a problem. By the way the "ALLOW FILTERING" is needed because you didn't specify the partition key in the request - add it to and you won't need "ALLOW FILTERING".

The "timestamp" column type (see documentation in https://docs.scylladb.com/stable/cql/types.html#timestamps) has millisecond resolution, so it is not surprising that attempting 6 significant digits returns an error.

There are a lot of remaining questions in this issue which are worthy writing tests for (test/cql-pytest/test_timestamp.py is a good place to put them) to verify what Scylla and Cassandra do. Among them:

1. Since a "timestamp" column has millisecond precision, it should print 3 digits of precision, not 6 (as Guy now reported) and not 0 (as reported in the original issue).
2. Converting a string constant to "timestamp" column should work, with different variations - with and without timezone specification, with and without milliseconds. Too many precision digits can generate an error.
3. Equality and inequality testing with properly-specified timestamp (with millisecond precision) should work properly.
4. The original issue was not about any of this stuff - it was about **arithmetic operations**, like the addition in `WHERE t >= toTimeStamp(now()) + 1d`. We need to figure out why @larisau thought that this syntax _should_ work. Is it supported by Cassandra?

After we write all these tests and figure out what is the current situation, we can open new issues and specific bugs and close this vague issue.

**Comment by nyh** [comment]
Answering some of my questions above using small tests which I wrote:

1. The CQL protocol returns a timestamp column as an integer, not the string "2023-01-03 07:03:28.253000+0000". What prints this string is cqlsh, which has `DEFAULT_TIMESTAMP_FORMAT = '%Y-%m-%d %H:%M:%S.%f%z'`. Cqlsh uses its `convert_datetime()` function to convert the millisecond count to a Python date_time, and then prints that using the above format. If we don't like this (and I don't) we need to create a cqlsh issue - it's not a Scylla issue.
2. I checked using too many digits of precision in a timestamp string constant. In Cassandra the extra digits are silently truncated, in Scylla an error is generated (`marshaling error: unable to parse date '2011-02-03 04:05:12.345678+0000': marshaling error: Milliseconds length exceeds expected (6)`). I don't think difference in error-handling is worth fixing, although I don't know why we decided to add this error in commit 633a13f7b3 (and why it got such obscure wording in commit 61b1988aa153160fea55019a001e9458882fe6bf).
3. I confirmed that converting a string to a timestamp behaves exactly as documented (and like Cassandra) for all documented variations of the format (except the error-handling of the over-precise time mentioned above).
4. I confirmed that various comparison operators (=, <, >, <=, >=) work as expected on timestamps.
5. I confirmed that the "arithmetic operations", which started this entire thread, indeed works on Cassandra 4 but doesn't on Scylla. This is a real issue, but a duplicate of #2693 and #2694.

All in all, this issue should be closed as a duplicate of #2693 and #2694. Most of the other discussion above is red herrings, and I'll send a test patch to verify that Scylla is working correctly except the unimplemented Cassandra 4.0 feature.
